Page cache spilling

el-yawd opened the pull request:
Closes #1661, #1625 and #2074

This PR also adds pragma `cache_spill` to manage cache spilling.

About #2074, I tested it locally and it works rather fine it only takes forever, about 10min in my laptop. We can turn spilling more efficient in future.

jussisaurio commented:
I closed this #1493 yesterday because with our temporary page cache size bump it works, so I tried to run that query on this branch and decreasing the page cache size back to 2000 (it's temporarily 100000)

```sql
turso> select
        c_count,
        count(*) as custdist
from
        (
                select
                        c_custkey,
                        count(o_orderkey) as c_count
                from
                        customer left outer join orders on
                                c_custkey = o_custkey
                                and o_comment not like '%express%packages%'
                group by
                        c_custkey
        ) as c_orders
group by
        c_count
order by
        custdist desc,
        c_count desc;
  × Internal error: Failed to insert loaded page 1247 into cache: Full { should_spill: true }
```

jussisaurio commented:
Can confirm that the mechanism works for non-ephemeral wal spilling

```
Turso v0.1.3-pre.3
Enter ".help" for usage hints.
This software is ALPHA, only use for development, testing, and experimentation.
turso> CREATE TABLE t(x);
turso> INSERT INTO t VALUES (randomblob(512 * 1024)); 
turso> INSERT INTO t SELECT * FROM t;
turso> INSERT INTO t SELECT * FROM t;
turso> INSERT INTO t SELECT * FROM t;
turso> INSERT INTO t SELECT * FROM t;
turso> INSERT INTO t SELECT * FROM t;
```

However as you mentioned, it is insanely slow even in release mode.

jussisaurio commented on core/storage/pager.rs:
```diff
@@ -1366,9 +1441,68 @@ impl Pager {
+                            "We somehow have a dirty page that isn't in the cache: page_id({page_id})",
+                        )
+                    });
+                    assert!(page.is_dirty());
```
`if page.is_pinned()` spilling it should be ignored

jussisaurio commented on core/storage/pager.rs:
```diff
@@ -1366,9 +1441,68 @@ impl Pager {
+                let in_flight = *self.flush_info.borrow().in_flight_writes.borrow();
+                if in_flight == 0 {
+                    self.flush_info.borrow_mut().state = CacheFlushState::Start;
+                    self.wal.borrow_mut().finish_append_frames_commit()?;
```
We shouldn't call `commit` here

el-yawd commented:
@jussisaurio addressed your comments and fixed some issues. Now using the `INSERT INTO` example it only panics due `OpenEphemeral` issue. Do you know any case that could trigger an cache full that doesn't use `OpenEphemeral`?

jussisaurio commented:
> @jussisaurio addressed your comments and fixed some issues. Now it only panics due `OpenEphemeral` issue. Do you know any case that could trigger cache full and doesn't use `OpenEphemeral`?

yes, this: https://github.com/tursodatabase/turso/pull/2210#issuecomment-3101279966

jussisaurio commented:
> <img alt="image" width="648" height="590" src="https://private-user-images.githubusercontent.com/79876389/469558959-04e320e4-6f64-44f1-9a71-b9c368a0f0dd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMyNDgyNjgsIm5iZiI6MTc1MzI0Nzk2OCwicGF0aCI6Ii83OTg3NjM4OS80Njk1NTg5NTktMDRlMzIwZTQtNmY2NC00NGYxLTlhNzEtYjljMzY4YTBmMGRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzIzVDA1MTkyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU1ZjM1MDAyYmUyMjRjYzUzMzhkMTc3NTM4Mzg1MjVlYjQ3MjQxM2NlMjY4ODVhNjEwZTJmZjNjNWVkZTM0NmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.TXxxu0hfoPzTeCeQB2F23mRBcWSLwTzIgp0Zqm9IDiA">

ah right, I forgot we have to do that for correctness

jussisaurio commented:
How about this one:

```sql
turso> CREATE TABLE t(x);
turso> INSERT INTO t VALUES (randomblob(1024 * 1024 * 30)); 
Done! Current db_size: 2001
Done! Current db_size: 3999
Done! Current db_size: 5997

thread 'main' panicked at core/storage/btree.rs:812:9:
we can't have more pages to read while also have read everything
```
